# Import data

In [33]:
import pandas as pd

#set option to view all columns
pd.set_option('display.max_columns', None)

#read data
data = pd.read_csv("Interview.csv")

data.shape

(1234, 28)

---
# View data

In [22]:
data.head(5)

Date of Interview Client name         Industry Location  \
0        13.02.2015     Hospira  Pharmaceuticals  Chennai   
1        13.02.2015     Hospira  Pharmaceuticals  Chennai   
2        13.02.2015     Hospira  Pharmaceuticals  Chennai   
3        13.02.2015     Hospira  Pharmaceuticals  Chennai   
4        13.02.2015     Hospira  Pharmaceuticals  Chennai   

  Position to be closed Nature of Skillset    Interview Type Name(Cand ID)  \
0   Production- Sterile            Routine  Scheduled Walkin   Candidate 1   
1   Production- Sterile            Routine  Scheduled Walkin   Candidate 2   
2   Production- Sterile            Routine  Scheduled Walkin   Candidate 3   
3   Production- Sterile            Routine  Scheduled Walkin   Candidate 4   
4   Production- Sterile            Routine  Scheduled Walkin   Candidate 5   

  Gender Candidate Current Location Candidate Job Location Interview Venue  \
0   Male                    Chennai                  Hosur           Hosur   
1   Male                    Chennai              Bangalore           Hosur   
2   Male                    Chennai                Chennai           Hosur   
3   Male                    Chennai                Chennai           Hosur   
4   Male                    Chennai              Bangalore           Hosur   

  Candidate Native location  \
0                     Hosur   
1                    Trichy   
2                   Chennai   
3                   Chennai   
4                   Chennai   

  Have you obtained the necessary permission to start at the required time  \
0                                                Yes                         
1                                                Yes                         
2                                                NaN                         
3                                                Yes                         
4                                                Yes                         

  Hope there will be no unscheduled meetings  \
0                                        Yes   
1                                        Yes   
2                                         Na   
3                                        Yes   
4                                        Yes   

  Can I Call you three hours before the interview and follow up on your attendance for the interview  \
0                                                Yes                                                   
1                                                Yes                                                   
2                                                NaN                                                   
3                                                 No                                                   
4                                                Yes                                                   

  Can I have an alternative number/ desk number. I assure you that I will not trouble you too much  \
0                                                Yes                                                 
1                                                Yes                                                 
2                                                NaN                                                 
3                                                Yes                                                 
4                                                 No                                                 

  Have you taken a printout of your updated resume. Have you read the JD and understood the same  \
0                                                Yes                                               
1                                                Yes                                               
2                                                NaN                                               
3                                                 No                                               
4             

---
# Check statistics


In [24]:
data.describe()

Unnamed: 23  Unnamed: 24  Unnamed: 25  Unnamed: 26  Unnamed: 27
count          0.0          0.0          0.0          0.0          0.0
mean           NaN          NaN          NaN          NaN          NaN
std            NaN          NaN          NaN          NaN          NaN
min            NaN          NaN          NaN          NaN          NaN
25%            NaN          NaN          NaN          NaN          NaN
50%            NaN          NaN          NaN          NaN          NaN
75%            NaN          NaN          NaN          NaN          NaN
max            NaN          NaN          NaN          NaN          NaN

>Because some last columns is entirely missing data, we can drop it

In [37]:
data.drop(["Unnamed: 23","Unnamed: 24","Unnamed: 25","Unnamed: 26","Unnamed: 27"],axis=1, inplace =True)

---
# View data again

In [32]:
data.shape

(1234, 23)

---
# Rename columns

In [34]:
#view all columns
data.columns

Index(['Date of Interview', 'Client name', 'Industry', 'Location',
       'Position to be closed', 'Nature of Skillset', 'Interview Type',
       'Name(Cand ID)', 'Gender', 'Candidate Current Location',
       'Candidate Job Location', 'Interview Venue',
       'Candidate Native location',
       'Have you obtained the necessary permission to start at the required time',
       'Hope there will be no unscheduled meetings',
       'Can I Call you three hours before the interview and follow up on your attendance for the interview',
       'Can I have an alternative number/ desk number. I assure you that I will not trouble you too much',
       'Have you taken a printout of your updated resume. Have you read the JD and understood the same',
       'Are you clear with the venue details and the landmark.',
       'Has the call letter been shared', 'Expected Attendance',
       'Observed Attendance', 'Marital Status', 'Unnamed: 23', 'Unnamed: 24',
       'Unnamed: 25', 'Unnamed: 26', 'Unname

In [44]:
#create column name
columns = ['date','name','industry','job_location','position','skillset','type','candidate_id','gender','candidate_address','wanted_location','venue','nation','is_ready','is_unschedule_meeting','precall_3h','additional_number','understand_jd','know_location','letter_received','expected_attendance','observed_attendance','marital_status']

#assign to data
data.columns=columns

---
# Check statistics again

In [48]:
data.describe(include=['O'])

date                     name industry job_location position  \
count         1233                     1234     1233         1233     1233   
unique          96                       16        7           11        7   
top     06.02.2016  Standard Chartered Bank     BFSI      Chennai  Routine   
freq           220                      904      949          754     1023   

                          skillset               type   candidate_id gender  \
count                         1233               1233           1233   1233   
unique                          92                  6           1233      2   
top     JAVA/J2EE/Struts/Hibernate  Scheduled Walk In  Candidate 619   Male   
freq                           220                456              1    965   

       candidate_address wanted_location    venue   nation is_ready  \
count               1233            1233     1233     1233     1029   
unique                10               7        7       46        7   
top              Chennai         Chennai  Chennai  Chennai      Yes   
freq                 754             893      852      595      917   

       is_unschedule_meeting precall_3h additional_number understand_jd  \
count                    986        986               986           985   
unique                     7          5                 6             8   
top                      Yes        Yes               Yes           Yes   
freq                     949        951               936           940   

       know_location letter_received expected_attendance observed_attendance  \
count            985             988                1228                1233   
unique             7              12                   7                   8   
top              Yes             Yes                 Yes                 Yes   
freq             946             932                 882                 701   

       marital_status  
count            1233  
unique              2  
top            Single  
freq              767

In [50]:
data.describe(percentiles=[0.1,0.25,0.5,0.75,0.9,0.95,0.99])

date                     name industry job_location position  \
count         1233                     1234     1233         1233     1233   
unique          96                       16        7           11        7   
top     06.02.2016  Standard Chartered Bank     BFSI      Chennai  Routine   
freq           220                      904      949          754     1023   

                          skillset               type   candidate_id gender  \
count                         1233               1233           1233   1233   
unique                          92                  6           1233      2   
top     JAVA/J2EE/Struts/Hibernate  Scheduled Walk In  Candidate 619   Male   
freq                           220                456              1    965   

       candidate_address wanted_location    venue   nation is_ready  \
count               1233            1233     1233     1233     1029   
unique                10               7        7       46        7   
top              Chennai         Chennai  Chennai  Chennai      Yes   
freq                 754             893      852      595      917   

       is_unschedule_meeting precall_3h additional_number understand_jd  \
count                    986        986               986           985   
unique                     7          5                 6             8   
top                      Yes        Yes               Yes           Yes   
freq                     949        951               936           940   

       know_location letter_received expected_attendance observed_attendance  \
count            985             988                1228                1233   
unique             7              12                   7                   8   
top              Yes             Yes                 Yes                 Yes   
freq             946             932                 882                 701   

       marital_status  
count            1233  
unique              2  
top            Single  
freq              767

#### There is a row missing all data, there is 1234 rows, but 1233 appears in all columns

In [52]:
data.tail(5)

date    name         industry job_location position  \
1229  07.05.2016  Pfizer  Pharmaceuticals      Chennai    Niche   
1230  06.05.2016  Pfizer  Pharmaceuticals      Chennai    Niche   
1231  06.05.2016  Pfizer  Pharmaceuticals      Chennai    Niche   
1232  06.05.2016  Pfizer  Pharmaceuticals      Chennai    Niche   
1233         NaN                    NaN          NaN      NaN   

                skillset        type    candidate_id  gender  \
1229        Biosimiliars  Scheduled   Candidate 1230    Male   
1230        Biosimiliars  Scheduled   Candidate 1231    Male   
1231  generic drugs – RA  Scheduled   Candidate 1232    Male   
1232  generic drugs – RA  Scheduled   Candidate 1233  Female   
1233                 NaN         NaN             NaN     NaN   

     candidate_address wanted_location    venue     nation is_ready  \
1229           Chennai         Chennai  Chennai  Hyderabad      Yes   
1230           Chennai         Chennai  Chennai  Hyderabad      Yes   
1231           Chennai         Chennai  Chennai    Chennai      Yes   
1232           Chennai         Chennai  Chennai    Kolkata      NaN   
1233               NaN             NaN      NaN        NaN      NaN   

     is_unschedule_meeting precall_3h additional_number understand_jd  \
1229                   Yes        Yes               Yes           Yes   
1230                   Yes        Yes               Yes           Yes   
1231                   Yes        Yes               Yes           Yes   
1232                   NaN        NaN               NaN           NaN   
1233                   NaN        NaN               NaN           NaN   

     know_location letter_received expected_attendance observed_attendance  \
1229           Yes             Yes                 Yes                 Yes   
1230           Yes             Yes                 Yes                 Yes   
1231           Yes             Yes                 Yes                 Yes   
1232           NaN             NaN           Uncertain                 Yes   
1233           NaN             NaN                 NaN                 NaN   

     marital_status  
1229         Single  
1230        Married  
1231         Single  
1232         Single  
1233            NaN

#### Drop last row

In [68]:
#data.drop[1233]

data.iloc[1233]

date                     NaN
name                      
industry                 NaN
job_location             NaN
position                 NaN
skillset                 NaN
type                     NaN
candidate_id             NaN
gender                   NaN
candidate_address        NaN
wanted_location          NaN
venue                    NaN
nation                   NaN
is_ready                 NaN
is_unschedule_meeting    NaN
precall_3h               NaN
additional_number        NaN
understand_jd            NaN
know_location            NaN
letter_received          NaN
expected_attendance      NaN
observed_attendance      NaN
marital_status           NaN
Name: 1233, dtype: object

In [73]:
data.drop(data.index[1233], inplace=True)

IndexError: index 1233 is out of bounds for axis 0 with size 1233

---
#### Check data again

In [74]:
data.describe(include=['O'])

date                     name industry job_location position  \
count         1233                     1233     1233         1233     1233   
unique          96                       15        7           11        7   
top     06.02.2016  Standard Chartered Bank     BFSI      Chennai  Routine   
freq           220                      904      949          754     1023   

                          skillset               type   candidate_id gender  \
count                         1233               1233           1233   1233   
unique                          92                  6           1233      2   
top     JAVA/J2EE/Struts/Hibernate  Scheduled Walk In  Candidate 619   Male   
freq                           220                456              1    965   

       candidate_address wanted_location    venue   nation is_ready  \
count               1233            1233     1233     1233     1029   
unique                10               7        7       46        7   
top              Chennai         Chennai  Chennai  Chennai      Yes   
freq                 754             893      852      595      917   

       is_unschedule_meeting precall_3h additional_number understand_jd  \
count                    986        986               986           985   
unique                     7          5                 6             8   
top                      Yes        Yes               Yes           Yes   
freq                     949        951               936           940   

       know_location letter_received expected_attendance observed_attendance  \
count            985             988                1228                1233   
unique             7              12                   7                   8   
top              Yes             Yes                 Yes                 Yes   
freq             946             932                 882                 701   

       marital_status  
count            1233  
unique              2  
top            Single  
freq              767

In [75]:
data.tail(5)

date    name         industry job_location position  \
1228  07.05.2016  Pfizer  Pharmaceuticals      Chennai    Niche   
1229  07.05.2016  Pfizer  Pharmaceuticals      Chennai    Niche   
1230  06.05.2016  Pfizer  Pharmaceuticals      Chennai    Niche   
1231  06.05.2016  Pfizer  Pharmaceuticals      Chennai    Niche   
1232  06.05.2016  Pfizer  Pharmaceuticals      Chennai    Niche   

                skillset        type    candidate_id  gender  \
1228  generic drugs – RA  Scheduled   Candidate 1229    Male   
1229        Biosimiliars  Scheduled   Candidate 1230    Male   
1230        Biosimiliars  Scheduled   Candidate 1231    Male   
1231  generic drugs – RA  Scheduled   Candidate 1232    Male   
1232  generic drugs – RA  Scheduled   Candidate 1233  Female   

     candidate_address wanted_location    venue     nation is_ready  \
1228           Chennai         Chennai  Chennai  Bangalore      Yes   
1229           Chennai         Chennai  Chennai  Hyderabad      Yes   
1230           Chennai         Chennai  Chennai  Hyderabad      Yes   
1231           Chennai         Chennai  Chennai    Chennai      Yes   
1232           Chennai         Chennai  Chennai    Kolkata      NaN   

     is_unschedule_meeting precall_3h additional_number understand_jd  \
1228                   Yes        Yes               Yes           Yes   
1229                   Yes        Yes               Yes           Yes   
1230                   Yes        Yes               Yes           Yes   
1231                   Yes        Yes               Yes           Yes   
1232                   NaN        NaN               NaN           NaN   

     know_location letter_received expected_attendance observed_attendance  \
1228           Yes             Yes                 Yes                 Yes   
1229           Yes             Yes                 Yes                 Yes   
1230           Yes             Yes                 Yes                 Yes   
1231           Yes             Yes                 Yes                 Yes   
1232           NaN             NaN           Uncertain                 Yes   

     marital_status  
1228        Married  
1229         Single  
1230        Married  
1231         Single  
1232         Single

---
## Check distribution value  and replace wrong value

In [78]:
data.job_location.value_counts()

Chennai       754
Bangalore     292
chennai        86
Hyderabad      38
Gurgaon        33
Noida          15
- Cochin-       9
chennai         3
Gurgaonr        1
CHENNAI         1
Delhi           1
Name: job_location, dtype: int64

In [83]:
data.date.value_counts()  #su dung format date time

06.02.2016               220
04/12/16                  86
13.04.2016                84
25.06.2015                43
02.09.2016                43
21.4.16                   38
22.01.2016                34
07.05.2016                31
03.04.2015                30
14.01.2016                28
27.02.2016                26
26.02.2016                26
20.08.2016                25
05/11/2016                25
25.5.2016                 24
23.02.2016                22
02.12.2015                19
19.06.2015                18
19.03.16                  17
30.01.2016                16
09-01-2016                15
13.06.2016                15
30.1.16                   14
25-5-2016                 14
21.4.2016                 14
11.5.16                   14
22.03.2016                13
13/03/2015                13
25.05.16                  12
07.06.2016                12
                        ... 
25 – Apr-16                2
28.08.2016 & 09.30 AM      2
21/04/16                   2
20 Apr 16     

---
# todo: check and do for all values in ***all columns*** here........

---
# And process to solve NA value, example here is understand_jd column: there are 4 methods:
1. Remove (when %population is too small, or value is not effect to model)
1. Replace by 0 (if data type is number) **ONLY when there is no other option**
1. Replace by mean/median value (if data type is number), but **choose what** 
1. Check correlation between futures,build model base on other futures --> predict the missing value

In [85]:
data.understand_jd.value_counts

<bound method IndexOpsMixin.value_counts of 0       Yes
1       Yes
2       NaN
3        No
4       Yes
5       Yes
6       Yes
7       Yes
8       Yes
9       Yes
10      Yes
11      Yes
12      Yes
13      Yes
14      Yes
15      Yes
16      Yes
17      Yes
18      Yes
19      Yes
20      Yes
21      Yes
22      Yes
23      Yes
24       No
25      Yes
26      Yes
27       No
28      Yes
29      Yes
       ... 
1203    Yes
1204    Yes
1205    Yes
1206    Yes
1207    Yes
1208    Yes
1209    Yes
1210    Yes
1211    Yes
1212    Yes
1213    Yes
1214    Yes
1215    Yes
1216    Yes
1217    Yes
1218    Yes
1219    Yes
1220    Yes
1221    Yes
1222    Yes
1223    Yes
1224    Yes
1225    Yes
1226    Yes
1227    Yes
1228    Yes
1229    Yes
1230    Yes
1231    Yes
1232    NaN
Name: understand_jd, Length: 1233, dtype: object>

----
# Convert from text to number
### All futures need to convert to number before build model
1. Convert from text to number
1. Apply one hot encoding, or using <font color='red'> get dummy </font> (PANDAS) has same
1.  test color <span style="color:#FFFF00"> tetst </span>

---
# Check distribution by drawing